### PR TITLE
chore: use nyc over istanbul-cli

### PR DESCRIPTION
**Pull request by JaKXz**
👋 Hi! `istanbuljs` member here -

Thought I'd help out a little by simplifying your coverage commands and configuration :) 

Otherwise, the main win of bringing you over to the much better supported `nyc` + `babel-plugin-istanbul` is that you're now getting _more accurate coverage reports and stats!_

Please let me know if anything concerns you - very happy to answer any questions you might have. Thanks again for this wonderful library!

### CODE REVIEW

**Comment by ljharb** [comment]
Can you elaborate more on why we should change our coverage method? What's different about `nyc` and why should we prefer it over `istanbul`?

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,16 @@
+root = true
```
Comment (ljharb) [inline .editorconfig]: unless this file can be kept in sync with CI (ie, changes to it will cause misformatted files to fail) i'd prefer the file not be included.

File: test/components/CalendarDay_spec.jsx [hunk 2]
```diff
@@ -30,7 +30,7 @@ describe('CalendarDay', () => {
       expect(wrapper.text()).to.equal(lastOfMonth.format('D'));
     });
 
-    it('contains arbitrary content if renderDay is provided', () => {
+    it.skip('contains arbitrary content if renderDay is provided', () => {
```
Comment (ljharb) [inline test/components/CalendarDay_spec.jsx]: just noting this line will need reverting before merging

File: test/mocha.opts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+--compilers js:babel-register,jsx:babel-register
```
Comment (ljharb) [inline test/mocha.opts]: this is a good change regardless; i'd be stoked to merge it in a separate PR :-)

File: test/mocha.opts [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+--require airbnb-js-shims
+--require ./test/_helpers/ignoreSVGStrings.jsx
+--recursive
+test/**/*_spec.js test/**/*_spec.jsx test/_helpers/restoreSinonStubs.js
```
Comment (ljharb) [inline test/mocha.opts]: however, the glob part of this configuration should remain in package.json, not in the mocha opts.

**Comment by JaKXz** [comment]
> What's different about nyc and why should we prefer it over istanbul?

`nyc` is the successor to the istanbul cli. `nyc` uses istanbul's APIs under the hood and there have been some features & fixes since `istanbul@1.x-alpha` was published.

It's preferred over the istanbul cli because 
- the istanbul cli is no longer really maintained, or, better stated: future development is happening in `nyc`. 
- `babel-plugin-istanbul` + `nyc` provides [more accurate coverage of ES2015+ code](https://github.com/dtinth/babel-plugin-__coverage__#is-the-resulting-coverage-differ-from-istanbulisparta) [hence coverage numbers dropping, relatively signficantly].

File: .editorconfig (hunk 1 above)
Comment (JaKXz) [inline .editorconfig]: I have found [EditorConfig](http://editorconfig.org) to be very helpful as a maintainer because it takes care of nuisances with whitespace and line endings and newlines etc. behind the scenes. e.g. the IntelliJ platform comes with support built in and almost all popular editors have a simple installer. It's worth noting that [GitHub also has native support](https://github.com/editorconfig/editorconfig.github.com/pull/48).

If you still prefer it not be added, I will remove it.

Comment (ljharb) [inline .editorconfig]: I'd love to have it, but something in CI would need to enforce it (the linter typically does this for us). Do you know of anything that can do that?

File: package.json [hunk 5]
```diff
@@ -9,14 +9,11 @@
...
-    "cover:clean": "rimraf coverage",
-    "cover:check": "babel-node ./node_modules/istanbul/lib/cli check-coverage",
-    "cover:mocha": "babel-node ./node_modules/istanbul/lib/cli cover node_modules/mocha/bin/_mocha -- --compilers js:babel-register,jsx:babel-register --require airbnb-js-shims --recursive ./test/_helpers test/**/*",
+    "precover": "rimraf coverage",
+    "cover": "cross-env NODE_ENV=test nyc mocha",
     "lint": "eslint --ext .js,.jsx src test",
-    "mocha": "mocha --compilers js:babel-register,jsx:babel-register --require airbnb-js-shims --recursive ./test/_helpers",
     "pretest": "npm run --silent lint",
-    "tests-only": "npm run mocha --silent test",
+    "tests-only": "mocha --silent",
```
Comment (ljharb) [inline package.json]: I think the file path needs to be restored here?

File: .editorconfig (hunk 1 above)
Comment (JaKXz) [inline .editorconfig]: https://github.com/jedmao/eclint looks promising!

Comment (JaKXz) [inline .editorconfig]: But I don't feel like it's a strong enough win for you, if it must be enforced like this.

I will remove it from this PR since it's not really the main focus.

File: package.json (hunk 5 above)
Comment (JaKXz) [inline package.json]: This was a typo on my part - the `--recursive` flag handles the globs for us.

**Comment by JaKXz** [comment]
@ljharb mind elaborating on

> lockfiles are harmful in non-top-level-apps.

If you publish with a `files` array in `package.json` you're not adding any weight to the package as a dependency. I notice this package uses `.npmignore` and I missed adding it there, so that was my bad. Is there something I'm missing?

**Comment by ljharb** [comment]
`files` arrays are incredibly dangerous, for one, but npmignore is fine.

If I publish a shrinkwrap file, yarn ignores it. If I publish a yarn lock file, yarn and npm ignore it. In no case am I ensuring my users get a deterministic dep tree. However, I am ensuring that if an in-range update is published that breaks me, my own tests will merrily pass but my users will merrily break.

Only an app is in the position to truly know that dependencies should be frozen.

**Comment by JaKXz** [comment]
Curiouser and curiouser

> files arrays are incredibly dangerous, for one

Mind elaborating on that?

**Comment by ljharb** [comment]
@JaKXz i'm on a number of common slacks and on freenode IRC, but here's fine too.

Basically, the absolute worst case of forgetting to include a new file in `npmignore` is that "your package is a tiny bit too large". The worst case of forgetting to include a new file in "files" is that "your package is subtly broken, in a way that's almost impossible for you, the author, (or your CI tests) to determine". Overarching package size just *isn't a big problem* anyways, and the worst case is *so* much worse with an "inclusion list" approach, that the only safe approach is an "exclusion list".

**Comment by JaKXz** [comment]
I see. That's a reasonable approach to safety. 

Anyway, are there any other changes you want to see in this PR?

**Comment by majapw** [comment]
@ljharb what do we need to do to get this merged in?

File: test/mocha.opts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+--compilers js:babel-register,jsx:babel-register
+--require airbnb-js-shims
+--require ./test/_helpers/ignoreSVGStrings.jsx
```
Comment (ljharb) [inline test/mocha.opts]: This fails to include the file that restores sinon stubs; please revert this part and restore `npm run mocha`, and have that pass the file there. `mocha` doesn't work with global before/after handlers if they're `--require`d in.

**Comment by ljharb** [comment]
@JaKXz i've rebased this branch; but it seems like your previous command (`nyc mocha`) is a bit wonky; it doesn't seem to take the same path arguments as `mocha` does; and we prefer to use `npm run mocha` as our central source of truth for the mocha command. Separately, adding the path args to `mocha.opts` doesn't seem to work either.

Is this not something `nyc` supports?

**Comment by JaKXz** [comment]
@ljharb sorry for the long turnaround on this! I think I've addressed your concerns now. :)

File: package.json [hunk 7]
```diff
@@ -9,14 +9,12 @@
...
     "clean": "rimraf lib",
-    "cover": "npm run cover:clean && npm run cover:mocha && npm run cover:check",
-    "cover:clean": "rimraf coverage",
-    "cover:check": "babel-node ./node_modules/istanbul/lib/cli check-coverage",
-    "cover:mocha": "babel-node ./node_modules/istanbul/lib/cli cover node_modules/mocha/bin/_mocha -- --compilers js:babel-register,jsx:babel-register --require airbnb-js-shims --recursive ./test/_helpers test/**/*",
+    "precover": "rimraf coverage",
+    "cover": "cross-env NODE_ENV=test nyc npm run mocha",
     "lint": "eslint --ext .js,.jsx src test",
-    "mocha": "mocha --compilers js:babel-register,jsx:babel-register --require airbnb-js-shims --recursive ./test/_helpers",
+    "mocha": "mocha ./test/_helpers test/**/*_spec.js*",
```
Comment (ljharb) [inline package.json]: is the `_spec.js` part needed? Every JS file inside `test` should be considered a test, and double-running the helpers shouldn't be a problem :-/

File: package.json [hunk 8]
```diff
@@ -9,14 +9,12 @@
...
     "pretest": "npm run --silent lint",
-    "tests-only": "npm run mocha --silent test",
+    "tests-only": "npm run mocha -- --silent",
```
Comment (ljharb) [inline package.json]: the `--silent` was intended for `npm`, not `mocha` - this change passes it to `mocha` instead. Can this line be reverted?

File: .nycrc [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  "require": [
+    "babel-register"
+  ],
+  "reporter": [
+    "text",
+    "lcov"
+  ],
+  "all": true,
+  "check-coverage": true,
+  "statements": 89,
```
Comment (ljharb) [inline .nycrc]: just one final re-confirmation - are these numbers exactly in line with our current coverage? If not, can we bump them to be? :-)

**Comment by JaKXz** [comment]
@ljharb here are the original watermarks, and I've dropped the numbers in `.nycrc` to reflect the current coverage.

screen shot 2017-03-31 at 8 58 35 pm

File: package.json (hunk 8 above)
Comment (JaKXz) [inline package.json]: fixed.

File: package.json (hunk 7 above)
Comment (JaKXz) [inline package.json]: fixed.

**Comment by majapw** [comment]
Party! Why did coveralls coverage go down so significantly though?

**Comment by JaKXz** [comment]
Because the instrumentation is much more accurate now, and we are measuring against _all_ JS and jsx files :) which I don't think was happening before.

**Comment by majapw** [comment]
Exciting! Also, I think I need to write more tests. :P @ljharb is this good? Can we merge it in? I'll lower the coveralls check for now and admin priv merge it, if that's the case.

Thanks for all the hard work @JaKXz! We super appreciate it. :D

**Comment by ljharb** [comment]
I've just re-rebased and pushed up; if travis-ci passes (even if coveralls drops) then this is good to merge 🎉

**Comment by ljharb** [comment]
Coveralls passed which probably means the thresholds aren't high enough; but I'm going to merge this now!

**Comment by ljharb** [comment]
(I filed #417 to combine coverage across all test runs, instead of just on node 4.6)

**Comment by ljharb** [comment]
Thanks @JaKXz! 🎉

**Comment by majapw** [comment]
Thanks y'all! @ljharb, I lowered the threshold so all following prs
wouldn't be breaking, but yes, i agree that they are too low.
